File: LeetCode/69.sqrt-x.cpp
/*
 * @lc app=leetcode id=69 lang=cpp
 *
 * [69] Sqrt(x)
 *
 * https://leetcode.com/problems/sqrtx/description/
 *
 * algorithms
 * Easy (32.94%)
 * Likes:    1150
 * Dislikes: 1764
 * Total Accepted:    511.5K
 * Total Submissions: 1.5M
 * Testcase Example:  '4'
 *
 * Implement int sqrt(int x).
 * 
 * Compute and return the square root of x, where x is guaranteed to be a
 * non-negative integer.
 * 
 * Since the return type is an integer, the decimal digits are truncated and
 * only the integer part of the result is returned.
 * 
 * Example 1:
 * 
 * 
 * Input: 4
 * Output: 2
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: 8
 * Output: 2
 * Explanation: The square root of 8 is 2.82842..., and since 
 * the decimal part is truncated, 2 is returned.
 * 
 * 
 */

// @lc code=start
class SolutionA
{
public:
    int mySqrt(int x)
    {
        int left = 1, right = x;
        while (left <= right)
        {
            int mid = left + (right - left) / 2;
            if (mid <= x / mid && (mid + 1) > (x / (mid + 1)))
                return mid;
            if (mid < x / mid)
                left = mid + 1;
            else
                right = mid - 1;
        }
        return right;
    }
};

class Solution
{
public:
    int mySqrt(int x)
    {
        double ans = x;
        while (abs(ans * ans - x) > 1e-5)
        {
            ans = (ans + x / ans) / 2;
        }
        return ans;
    }
};
// @lc code=end

File: lc-archived/java/69.sqrt-x.java
/*
 * @lc app=leetcode id=69 lang=java
 *
 * [69] Sqrt(x)
 */

// @lc code=start
class Solution {
    public int mySqrt(int x) {
        double ans = x / 2.0;
        while (Math.abs(x - ans * ans) >= 1e-5) {
            ans = (ans + x / ans) / 2;
        }
        return (int) ans;
    }
}

// @Deprecated
// Failed for the case like 36 !!! since the answer may approximate as 5.99999999 and produce the answer as 5.

class Solution {

    public int mySqrt(int x) {
        double low = 1, high = x;
        while (low <= high) {
            double mid = (low + high) / 2;
            System.out.println(low + " " + mid + " " + high);
            if (Math.abs(x - mid * mid) < 1e-5)
                return (int) mid;
            if (x - mid * mid > 1e-5) {
                low = mid;
            } else {
                high = mid;
            }
        }
        return 0;
    }
}

// Note the difference between the commented line and the uncommented line.
class Solution {
    public int mySqrt(int x) {
        int low = 1, high = x;
        while (low <= high) {
            // int mid = (low + high) / 2;

            int mid = low + (high - low) / 2;

            // if (mid * mid <= x && (mid + 1) * (mid + 1) > x)
            if (mid <= x / mid && (mid + 1) > (x / (mid + 1)))
                return mid;
            // if (x - mid * mid > 0) {
            if (mid < x / mid) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return 0;
    }
}

// @lc code=end
